connectors/src/connectors/confluence/temporal/workflows.ts

```diff
@@ -0,0 +1,121 @@
+import type { ModelId } from "@dust-tt/types";
+import {
+  executeChild,
+  proxyActivities,
+  workflowInfo,
+} from "@temporalio/workflow";
+
+import type * as activities from "@connectors/connectors/confluence/temporal/activities";
+import type { DataSourceConfig } from "@connectors/types/data_source_config";
+
+// The Temporal bundle does not support the use of aliases in import statements.
+import { makeConfluenceSpaceSyncWorkflowIdFromParentId } from "./utils";
+
+const {
+  confluenceGetSpaceNameActivity,
+  confluenceListPageIdsInSpaceActivity,
+  confluenceSaveStartSyncActivity,
+  confluenceUpsertPageActivity,
+  confluenceSaveSuccessSyncActivity,
+
+  fetchConfluenceConfigurationActivity,
+  getSpaceIdsToSyncActivity,
+} = proxyActivities<typeof activities>({
+  startToCloseTimeout: "20 minutes",
+});
+
+export async function confluenceFullSyncWorkflow({
+  connectionId,
+  connectorId,
+  dataSourceConfig,
+  spaceIdsToBrowse,
+}: {
+  connectionId: string;
+  connectorId: ModelId;
+  dataSourceConfig: DataSourceConfig;
+  spaceIdsToBrowse?: string[];
+}) {
+  await confluenceSaveStartSyncActivity(connectorId);
+
+  const spaceIdsToSync =
+    spaceIdsToBrowse ?? (await getSpaceIdsToSyncActivity(connectorId));
+
+  const {
+    workflowId,
+    searchAttributes: parentSearchAttributes,
+    memo,
+  } = workflowInfo();
+
+  for (const spaceId of spaceIdsToSync) {
+    await executeChild(confluenceSpaceSyncWorkflow, {
+      workflowId: makeConfluenceSpaceSyncWorkflowIdFromParentId(
+        workflowId,
+        spaceId
+      ),
+      searchAttributes: parentSearchAttributes,
+      args: [
+        {
+          connectionId,
+          connectorId,
+          dataSourceConfig,
+          isBatchSync: true,
+          spaceId,
+        },
+      ],
+      memo,
+    });
+  }
+
+  await confluenceSaveSuccessSyncActivity(connectorId);
+}
+
+interface ConfluenceSpaceSyncWorkflowInput {
+  connectionId: string;
+  connectorId: ModelId;
+  dataSourceConfig: DataSourceConfig;
+  isBatchSync: boolean;
+  spaceId: string;
+}
+
+export async function confluenceSpaceSyncWorkflow(
+  params: ConfluenceSpaceSyncWorkflowInput
+) {
+  const uniquePageIds = new Set<string>();
+
+  const confluenceConfig = await fetchConfluenceConfigurationActivity(
+    params.connectorId
+  );
+
+  const { cloudId: confluenceCloudId } = confluenceConfig;
+
+  const spaceName = await confluenceGetSpaceNameActivity({
+    ...params,
+    confluenceCloudId: confluenceConfig?.cloudId,
+  });
+
+  // Retrieve and loop through all pages for a given space.
+  let nextPageCursor: string | null = "";
+  do {
+    const { pageIds, nextPageCursor: nextCursor } =
+      await confluenceListPageIdsInSpaceActivity({
+        ...params,
+        confluenceCloudId,
+        pageCursor: nextPageCursor,
+      });
+
+    nextPageCursor = nextCursor; // Prepare for the next iteration.
+
+    pageIds.forEach((id) => uniquePageIds.add(id));
```

Comment:
So here you accumulate in the workflow all the pages right?

If accurate, this is not super desirable as we want to avoid storing too much state in temporal.

One solution we have used in the past is to record these pages in DB, marked for update (maybe the same model maybe another one, up to you)

Then you can run an activity that is in charge of going a ahead and iterating locally on all pages to be udpated

Reply:
As per IRL let's lean on temporal to store that state :+1: